test/integration/puppetlabs/services/jruby/jruby_pool_int_test.clj

```diff
@@ -236,90 +237,149 @@
+           (jruby-protocol/return-instance jruby-service instance2 :max-requests-flush-while-pool-flush-in-progress-test)
+           (is (true? (check-jrubies-for-constant-counts pool-context 2 1))))
+
+         ;; now we'll set the ruby constant on the 3 instances in the new pool
+         (is (true? (set-constants-and-verify pool-context 3)))
+
+         ;; The flush should still not have completed at this point because
+         ;; the last instance from the old pool is still being borrowed.
+         ;; Wait for some arbitrary (but short) period of time to see that
+         ;; that the agent is still busy - presumably handling the flush.
+         (is (false? (await-for 50 pool-agent)))
+
+         ;; and finally, we return the last instance from the old pool
+         (jruby-protocol/return-instance jruby-service instance1 :max-requests-flush-while-pool-flush-in-progress-test)
+
+         ;; wait until the flush is complete
+         (is (true? (timed-await pool-agent))
+             (str "timed out waiting for the flush to complete, stack:\n"
+                  (get-all-stack-traces-as-str))))
+
+       ;; we should have three instances with the constant and one without.
+       (is (true? (check-jrubies-for-constant-counts pool-context 3 1)))
+
+       ;; The jruby return instance calls done within the previous
+       ;; check jrubies call may cause an instance to be in the process of
+       ;; being flushed when the server is shut down.  This ensures that
+       ;; the flushing is all done before the server is shut down - since
+       ;; that could otherwise cause an annoying error message about the
+       ;; pool not being full at shut down to be displayed.
+       (timed-await (:flush-instance-agent pool-context))))))
+
+(deftest initialization-and-cleanup-hooks-test
+  (testing "custom initialization and cleanup callbacks get called appropriately"
+    (let [lifecycle-fns {:initialize (fn [instance] (assoc instance :foo "FOO"))
+                         :shutdown (fn [instance] (log/error "Terminating" (:foo instance)))}
+          config (assoc-in (jruby-testutils/jruby-tk-config
+                            (jruby-testutils/jruby-config
+                             {:max-active-instances 1
+                              :max-requests-per-instance 10
+                              :borrow-timeout default-borrow-timeout}))
+                           [:jruby :lifecycle] lifecycle-fns)]
+      (logutils/with-test-logging
+       (tk-testutils/with-app-with-config
+        app
+        [jruby/jruby-pooled-service]
+        config
+        (let [jruby-service (tk-app/get-service app :JRubyService)
+              context (tk-services/service-context jruby-service)
+              pool-context (:pool-context context)]
+          ;; set a ruby constant in each instance so that we can recognize them
+          (is (true? (set-constants-and-verify pool-context 1)))
```

Comment:
For some reason I though that this was necessary in order to verify that the instance(s) were up before trying to borrow, but I guess that's not actually true, and you're right that we're not calling the corresponding `verify-no-constants` so this shouldn't be necessary.